Propagate panics across joins instead of hanging

Amanieu opened the pull request:
Fixes #10

nikomatsakis commented:
Nice, thanks! This seems like a good start, but I don't think it's quite there. For example, if a panic occurs here, it will propagate all the way up and kill the worker thread -- but since it doesn't affect the core thread pool, it means we'll just continue with one less thread.

What I had in mind for a stopgap measure was to add another guard into the worker thread that will poison the central thread pool. Then at various points (e.g., when a job is stolen, or when a new job is submitted from a non-worker-thread) we can check for poisoning and propagate errors. This will at least eagerly bring rayon down once a panic occurs. It's not great, but it's better than hanging.

Another alternative might be to send a message to the main pool that will cause it to restart the thread instead, so that some jobs go down, but eventually the pool recovers. This would probably be easier/better to do based on the `recover` API.

Amanieu commented:
I've added a second guard to terminate the thread pool if a worker thread panics. I also fixed the handling of terminate to properly wake up all threads.

I'm planning on implementing proper panic propagation based on `recover` behind a `cfg(nightly)` switch. This would send panics back the caller of `join`, but it will always keep the pool alive.

nikomatsakis commented:
@Amanieu 

> I've added a second guard to terminate the thread pool if a worker thread panics. I also fixed the handling of terminate to properly wake up all threads.

Great! I'll take a look shortly.

> I'm planning on implementing proper panic propagation based on recover behind a cfg(nightly) switch. This would send panics back the caller of join, but it will always keep the pool alive.

Even better :)

nikomatsakis commented on src/test.rs:
```diff
@@ -107,3 +107,9 @@ fn positive_test_run_pass() {
+#[test]
+#[should_panic]
+fn panic() {
+    join(|| {}, || panic!("should panic"));
```
I am vaguely concerned that having a panic in a test like this will infect the global schedule and affect other tests -- we may want to move these kinds of panic-y tests to run-fail tests using the `compiletest` infrastructure. We'd have to kind of copy-and-paste the `positive_tests_run_pass` test right above and change the `mode` variable to `fun-fail` and the directory to `neg-tests-run`. Then you can make `.rs` files that are expected to panic. I think they have to have a comment like `// error-pattern: foo`, where `foo` is a string we expected to see in the output.

Amanieu replied:
Oops, I thought I removed that test. Yes it does have undesirable effects on the other tests. I'll look into adding some run-fail tests.

nikomatsakis commented on src/thread_pool.rs:
```diff
@@ -19,10 +18,10 @@ pub struct Registry {
 }
 
 struct RegistryState {
+    terminate: bool,
```
Nice, I like adding this here. Ensures we will see it eventually, but it doesn't interfere with the hot paths.

nikomatsakis commented on src/thread_pool.rs:
```diff
@@ -112,11 +116,14 @@ impl Registry {
         log!(InjectJobs { count: injected_jobs.len() });
         let mut state = self.state.lock().unwrap();
+        if state.terminate {
+            panic!("attempted to inject a job after thread pool shut down");
```
This error message will be seen by end-users, right? That is, if a panic occurs in one place, it would get propagated to later jobs?

Perhaps we should make it a bit more readable, something like: 

"rayon thread pool is contaminated by a previous panic; recovery is currently unimplemented, see <https://github.com/nikomatsakis/rayon/issues/10> for details"

nikomatsakis commented:
OK, this looks nice. The main thing that would be nice is if you could adapt the tests to use `run-fail` and add a few more detailed tests. For example:

0. The existing "shallow panic" test you wrote.
1. Use par_iter with `weight_max` (to ensure fine-grained jobs) and map over a large array, and then panic in one of the maps. Something like `(0..2_000_000).into_par_iter().weight_max().foreach(|i| if i == 1_350_000 { panic!("boom") })`. I would of course expect this to propagate.
2. Start one thread that does a job and panics while using rayon. After it finishes, start another that does the same. The panic should propagate.

I guess that's all I can think of. :)

Amanieu commented:
I've pushed an updated version that fixes your comments. However there is still a big issue: the `iter_panic.rs` test sometimes segfaults, but I haven't been able to figure out the cause. It seems this is due to a thread having pointers into the stack of another thread while that thread is being unwound.

nikomatsakis commented:
@Amanieu hmm, yes, that makes sense. We will probably want to add some guards that ensure that we block on latches when unwinding. Can't think about this just now though.

An alternative of course is to stop using the stack, but I'm pretty reluctant to do that.

Amanieu commented:
I've refactored the code that deals with job objects, which seems to have fixed the segfault. It should also slightly improve benchmark numbers :)

I've also added panic propagation under the "nightly" cargo feature.

nikomatsakis commented:
@Amanieu awesome, sorry I havent' had a chance to take a look yet :) I'll try to do it today!

nikomatsakis commented on src/job.rs:
```diff
@@ -1,56 +1,99 @@
+    fn run_result(func: F) -> JobResult<R> {
+        // RecoverSafe is a bit ugly to deal with...
+        // Ideally this would just be panic::recover(func)
+        let mut wrapper = AssertRecoverSafe::new(Some(func));
```
It'd be good to add a comment explaining why we think it's legit to call this function "recover safe" despite having no idea what kind of data it touches. (The answer of course is because we propagate panics.) So perhaps just write: "// We can call this recover safe because we propagate panics."

nikomatsakis commented on src/job.rs:
```diff
@@ -1,56 +1,99 @@
+        // RecoverSafe is a bit ugly to deal with...
+        // Ideally this would just be panic::recover(func)
+        let mut wrapper = AssertRecoverSafe::new(Some(func));
+        match panic::recover(move || (*wrapper).take().unwrap()()) {
```
Can you add `FIXME(rust-lang/rust#XXX)` where XXX is the number of your PR?

nikomatsakis commented on src/thread_pool.rs:
```diff
@@ -109,14 +113,18 @@ impl Registry {
     }
 
-    pub unsafe fn inject(&self, injected_jobs: &[*mut Job<LockLatch>]) {
+    pub unsafe fn inject(&self, injected_jobs: &[*mut (Job<LockLatch> + 'static)]) {
```
Is the explicit `'static` needed here?

Amanieu replied:
Yes, it fails to compile otherwise. It's automatically added by lifetime inference in other places, but it needs to be explicitly specified here.

nikomatsakis replied:
Huh. I'm a bit surprised, but I guess I forget what the defaulting rules are around `*mut`.

Amanieu replied:
It's because it's defaulting to the lifetime of the outer `&[]`.

nikomatsakis replied:
Yes, I just expected it to be overridden by the `*mut`.

nikomatsakis commented:
So, as I said on IRC, this looks great!

In later PRs, it'd be nice to play with the setup of `Job` and see if we can't go back to using a trait somehow (or maybe we'll eventually fix stuff like `self: *mut Self`...), and of course it'd be nice to write more exhaustiv tests, but this is obviously far better than what's there now.

cuviper commented:
Nice work, @Amanieu, thanks!
